lambda/lambda.mli

```diff
@@ -603,12 +603,40 @@ type let_kind = Strict | Alias | StrictOpt
       we can discard e if x does not appear in e'
  *)
 
+(** A unique barrier is necessary on all references to a value that
+    do not count as a use in the uniqueness analysis. If that happens,
+    the relevant reference in typedtree is annotated with a [Unique_barrier.t].
+    Currently, these references are mapped to:
+    - [Texp_field] ~> [Pfield]+co
+    - [Texp_record] with an extended expression ~> [Pfield]+co, [Pduprecord]
+    - Pattern matches ~> [Pfield]+co, [Parrayrefu], [Punboxed_product_field],
+        [Parraylength], [Lstringswitch]
+    where [Pfield]+co also includes [Pfloatfield], [Pufloatfield], [Pmixedfield].
+
+    If the unique_barrier is set in the typedtree, the generated lambda code
+    may not be pushed down. We achieve this as follows:
+    - [Pfield]+co: we use [add_barrier_to_read]
+    - [Punboxed_product_field]: an unboxed product is not an allocation,
+        so no protection is necessary
+    - [Pduprecord]: this one can not be pushed down
+    - [Parrayrefu]: we use [add_barrier_to_mutable_flag]
+    - [Parraylength]: TODO this one is currently unsafe
+    - [Lstringswitch]: we do not mind since strings mode-cross uniqueness
+
+    There are many other primitives in lambda that can read from an allocation,
+    such as: [Parrayrefs], [Pduparray], [Pbyteslength], [Pbytesrefu],
+    [Pbytesrefs], [Pbigarrayref], [Pbigarraydim], [Pbytes_load*],
+    [P.*_array_load_128], [Patomic_load], [Patomic_exchange], [Patomic_cas],
+    [Patomic_fetch_add], [Punbox_vector], [Pbox_vector].
+    However, none of these can be generated by a one of the typedtree nodes. *)
```

Comment:
> one of the typedtree nodes

You mean the ones above? I think there's a slightly more direct argument: These primitives are produced only by syntax that counts as a unique use. Is that right?